lnwallet/channel_test.go

```diff
@@ -3870,3 +3870,53 @@ func TestSignCommitmentFailNotLockedIn(t *testing.T) {
 }
 
 // TODO(roasbeef): testing.Quick test case for retrans!!!
+
+// TestDesyncHTLCs checks that we cannot add HTLCs that would make the
+// balance negative, when the remote and locak update logs are desynced.
+func TestDesyncHTLCs(t *testing.T) {
+	t.Parallel()
+
+	// We'll kick off the test by creating our channels which both are
+	// loaded with 5 BTC each.
+	aliceChannel, bobChannel, cleanUp, err := createTestChannels(1)
+	if err != nil {
+		t.Fatalf("unable to create test channels: %v", err)
+	}
+	defer cleanUp()
+
+	// First add one HTLC of value 4.1 BTC.
+	htlcAmt := lnwire.NewMSatFromSatoshis(4.1 * btcutil.SatoshiPerBitcoin)
+	htlc, preimage := createHTLC(0, htlcAmt)
+	if _, err := aliceChannel.AddHTLC(htlc); err != nil {
+		t.Fatalf("unable to add htlc: %v", err)
+	}
+	if _, err := bobChannel.ReceiveHTLC(htlc); err != nil {
+		t.Fatalf("unable to recv htlc: %v", err)
+	}
+
+	// Now let let Bob fail this HTLC.
```

Comment:
This is invalid (already). Bob *cannot* fail this HTLC as it isn't yet locked in.

Reply:
You are right, found the protocol requirement here: https://github.com/lightningnetwork/lightning-rfc/blob/master/02-peer-protocol.md#requirements-10

However, I added a `forceStateTransition` here, which should lock it in. Test triggers the same negative balance.